### PR TITLE
Add checks for code quality: usage of tf.compat.v1 and experimental APIs

### LINKED ISSUE #1279: Add tests to ensure that we don't use experimental/private/deprecated tensorflow APIs
**Describe the feature and the current behavior/state.**

Everything is in the title. It's hard to monitor this. Having a test in the CI would help us make sure we don't start using private/experimental APIs. The more we use them, the harder it's going to be when we bump tensorflow versions. It's also a requirements if we want to be able to support multiple versions of tensorflow at the same time (I know, crazy talk, right?).

In the long run, that's one more check that we won't have to do manually as maintainers.

We should also check for `tf.compat.v1`. If there is a `tf.compat.v1` that we can't remove, it means that we should ask the TF team that they add an API for an equivalent function.

**Relevant information**
- Are you willing to contribute it (yes/no): yes
- Are you willing to maintain it going forward? (yes/no): yes, we might even need it for keras
- Is there a relevant academic paper? (if so, where): no
- Is there already an implementation in another framework? (if so, where): not to my knowledge
- Was it part of tf.contrib? (if so, where):no 

**Which API type would this fall under (layer, metric, optimizer, etc.)**

Tests

**Who will benefit with this feature?**

Us devs, and users because we'll be able to support more TF versions.

**Any other info.**

I've no idea about how to implement that cleanly. (analysing the AST maybe?) Ideas welcome.

### PR DESCRIPTION
**Pull request by gabrieldemarmiesse**
This should save us some work during code reviews. 

We might want to discuss for this policy about experimental APIs though. See https://github.com/tensorflow/community/pull/218
Should we be more permissive? Less permissive?

Closes  #1279

### CODE REVIEW

**Comment by AakashKumarNain** [comment]
IMO we should definitely block anything that uses `tf.compat.v1`. If we allow this, we would be in somewhat similar scenario as `py2` -> `py3`. 

Regarding `experiemtal`, although most of the time things remain the same when they come out of experimental policy, IMO we should still defer its usage.

**Comment by gabrieldemarmiesse** [comment]
The reason behing disallowing experimental in user-facing code is to be forward compatible. In a perfect world, since the public API of TF is backward compatible (with the exception of experimental APIs) if we use only the public API, the users should be able to use any version of TF above the recommended one and it will work. 

Experimental API are doomed to have their name changed as some points, which means that we they do, it forces users to update their TF addons version. Users should have to think as little as possible about compatibility between packages.

**Comment by AakashKumarNain** [comment]
I am fine with the changes. @seanpmorgan thoughts?
